drivers/adc/adc_ad4114.c

```diff
@@ -0,0 +1,457 @@
+	struct k_thread thread;
+	struct k_sem sem;
+	uint16_t channels;
+	uint16_t channels_cfg;
+	uint32_t *buffer;
+	uint32_t *repeat_buffer;
+
+	K_KERNEL_STACK_MEMBER(stack, CONFIG_ADC_AD4114_ACQUISITION_THREAD_STACK_SIZE);
+};
+
+static int ad4114_write_reg(const struct device *dev, enum ad4114_reg reg_addr, uint8_t *buffer,
+			    size_t reg_size)
+{
+	int ret;
+	const struct adc_ad4114_config *config = dev->config;
+	uint8_t buffer_tx[5] = {0}; // One byte command, max 4 bytes data
+
+	const struct spi_buf tx_buf[] = {{
+		.buf = buffer_tx,
+		.len = ARRAY_SIZE(buffer_tx),
+	}};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf),
+	};
+
+	buffer_tx[0] = AD4114_CMD_WRITE | reg_addr;
+
+	if (reg_size > 4) {
+		LOG_ERR("Invalid size, max data write size is 4");
+		return -ENOMEM;
+	}
+	// Fill the data
+	for (uint8_t index = 0; index < reg_size; index++) {
+		buffer_tx[1 + index] = buffer[index];
+	}
+
+	ret = spi_write_dt(&config->spi, &tx);
+	if (ret != 0) {
+		LOG_ERR("Write reg : error writing register 0x%X (%d)", reg_addr, ret);
+		return ret;
+	}
+
+	return ret;
+}
+
+static int ad4114_read_reg(const struct device *dev, enum ad4114_reg reg_addr, uint8_t *buffer,
+			   size_t reg_size)
+{
+	int ret;
+	const struct adc_ad4114_config *config = dev->config;
+
+	uint8_t buffer_tx[6] = {0};
+	uint8_t buffer_rx[ARRAY_SIZE(buffer_tx)] = {0xFF};
+	const struct spi_buf tx_buf[] = {{
+		.buf = buffer_tx,
+		.len = ARRAY_SIZE(buffer_tx),
+	}};
+	const struct spi_buf rx_buf[] = {{
+		.buf = buffer_rx,
+		.len = ARRAY_SIZE(buffer_rx),
+	}};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf),
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = ARRAY_SIZE(rx_buf),
+	};
+	buffer_tx[0] = AD4114_CMD_READ | reg_addr;
+
+	ret = spi_transceive_dt(&config->spi, &tx, &rx);
+	if (ret != 0) {
+		LOG_ERR("Read reg: error writing register 0x%X (%d)", reg_addr, ret);
+		return ret;
+	}
+
+	// Copy received data in output buffer
+	for (uint8_t index = 0; index < reg_size; index++) {
+		buffer[index] = buffer_rx[index + 1];
+	}
+
+	return ret;
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_ad4114_data *data = CONTAINER_OF(ctx, struct adc_ad4114_data, ctx);
+
+	data->channels = ctx->sequence.channels;
+	data->repeat_buffer = data->buffer;
+
+	k_sem_give(&data->sem);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx, bool repeat_sampling)
+{
+	struct adc_ad4114_data *data = CONTAINER_OF(ctx, struct adc_ad4114_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int adc_ad4114x_validate_buffer_size(const struct device *dev,
+					    const struct adc_sequence *sequence)
+{
+	uint8_t channels;
+	size_t needed;
+
+	channels = POPCOUNT(sequence->channels);
+	needed = channels * sizeof(uint32_t);
+
+	if (sequence->buffer_size < needed) {
+		LOG_ERR("insufficient buffer size");
```

Comment:
This message is also being printed in line 228.

Reply:
I only keep the log in `adc_ad4114_start_read` function